test/matgen.hh

```diff
@@ -0,0 +1,40 @@
+// Copyright (c) 2017-2023, University of Tennessee. All rights reserved.
+// SPDX-License-Identifier: BSD-3-Clause
+// This program is free software: you can redistribute it and/or modify it under
+// the terms of the BSD 3-Clause license. See the accompanying LICENSE file.
+
+
+#ifndef SLATE_GENERATE_MATGEN_HH
+#define SLATE_GENERATE_MATGEN_HH
+
+
+#include "test.hh"
+#include "../matgen/matgen_params.hh"
+
+namespace slate {
+
+template <typename matrix_type>
+void generate_matrix(
+    MatrixParams& params,
+    matrix_type& A,
+    slate::Options const& opts = slate::Options() )
+{
+    MatgenParams mg_params;
+    mg_params.kind = params.kind();
+    mg_params.cond_request = params.cond_request();
+    mg_params.condD = params.condD();
+    mg_params.seed = params.seed();
+    mg_params.marked = params.marked();
+
+    //mg_params.generate_label() = params.generate_label();
+
+    generate_matrix( mg_params, A, opts);
+
+    //label will be put here as output as well
+    mg_params.cond_actual = params.cond_actual();
```

Comment:
This should be `params.cond_actual() = mg_params.cond_actual` since `cond_actual` is an output variable.

Also, I believe after this line is were to place
```C++
if (params.marked())
        params.generate_label();
```

Reply:
@neil-lindquist Thank you for pointing that out to me. Could you clarify what you mean by the second section? I believe you are wanting me to add `params.cond_actual();` to the following code block:

```
if (params.marked())
        params.generate_label();
        params.cond_actual();
```

However, I may be misunderstanding.

Reply:
My suggestion was to replace lines 33 and 34 linked w/ this comment chain with
```C++
params.cond_actual() = mg_params.cond_actual;
if (params.marked())
        params.generate_label();
```
`params.cond_actual()` "returns" `double&`, so we can assign values to it (and need to if we want to push the value of `cond_actual` from the `matgen` layer to the `tester` layer).

(Side C++ advice, when an if statement doesn't use `{}`, only the first statement will be conditional.  So, your `cond_actual` line isn't wrapped by the if, but it's indented as if it were.)

Reply:
I see, thanks for the clarification. I have added `params.cond_actual() = mg_params.cond_actual;`. However, I do not believe the following section is needed: 
```
if (params.marked())
        params.generate_label();
```
The reason is that within `generate_matrix_utils.cc` the function `decode_matrix( ... )` has this exact line of code. Within both `generate_matrix_ge.cc` and `generate_matrix_he_and_tz.cc`, `decode_matrix( ... )` is called. 

Therefore, I believe we only need to add: `params.label() = mg_params.label;` to `matgen.hh`.  I have tested this and it seems to index correctly to me. Test case below after adding `params.label() = mg_params.label;`:

```
% input: ./tester --dim 1k,2k --matrix rand,randn,circul gemm
% 2023-12-21 16:01:07, 1 MPI ranks, CPU-only MPI, 80 OpenMP threads, 8 GPU devices per MPI rank
                                                                                                                                                                                                         
type  origin  target  gemm   go   A   B   C   transA   transB       m       n       k      alpha       beta    nb    p    q  la      error   time (s)       gflop/s  ref time (s)   ref gflop/s  status  
   d    host    task  auto  col   1   1   1  notrans  notrans    1000    1000    1000   3.1+1.4i   2.7+1.7i   384    1    1   1   2.19e-16     0.0213        93.870            NA            NA  pass    
   d    host    task  auto  col   1   1   1  notrans  notrans    2000    2000    2000   3.1+1.4i   2.7+1.7i   384    1    1   1   1.80e-16     0.0565       283.209            NA            NA  pass    
   d    host    task  auto  col   2   1   1  notrans  notrans    1000    1000    1000   3.1+1.4i   2.7+1.7i   384    1    1   1   5.59e-16     0.0171       116.945            NA            NA  pass    
   d    host    task  auto  col   2   1   1  notrans  notrans    2000    2000    2000   3.1+1.4i   2.7+1.7i   384    1    1   1   5.30e-16     0.0411       389.127            NA            NA  pass    
   d    host    task  auto  col   3   1   1  notrans  notrans    1000    1000    1000   3.1+1.4i   2.7+1.7i   384    1    1   1   2.30e-16     0.0140       142.934            NA            NA  pass    
   d    host    task  auto  col   3   1   1  notrans  notrans    2000    2000    2000   3.1+1.4i   2.7+1.7i   384    1    1   1   1.83e-16     0.0401       399.208            NA            NA  pass    

% Matrix kinds:
%  1: rand, cond unknown
%  2: randn, cond unknown
%  3: circul, cond unknown

% All tests passed: gemm
```

Let me know what you think. Also, thank you for the C++ comment, I remember reading that a bit ago, but did forget about that nuance until you just mentioned it again.

Reply:
That can work.

Although, I'd personally argue that the `marked` semantics shouldn't be part of `libmatgen` since it's recording whether a `params` object is controlled by `tester`'s CLI.  Also, `generate_label` is just how we pretty-print the `params` object, so it feels like it belongs to the tester's UI, not to the matrix generator.

Reply:
I moved labels back to test directory.